Tutorial for memcached-operator with oria-operator (#64)
operator-framework/oria-operator: docs/tutorial.md

# Oria Operator Tutorial

An in-depth walkthrough of building and running a oria-operator with memcached-operator.

## Prerequisites

- [Operator SDK](https://sdk.operatorframework.io/docs/installation/) v1.21.0 or newer
- User authorized with `cluster-admin` permissions so that you can install an operator, CRDs and change rbac.
- [GNU Make](https://www.gnu.org/software/make/)

> 

Reviewer: This might be a bit overkill, but I think it would be nice to add here that you have to have a Kubernetes cluster running to do this stuff on. I think it would be good to include information about [KinD](https://kind.sigs.k8s.io/) in case they want to use a throwaway cluster for testing instead of one that they already have set up.

Author: Updated to Create an [Kind Cluster](https://kind.sigs.k8s.io/)

## Overview

We will create a sample project to let you know how it works and this sample will:

- Create a `memcached-operator` using below mentioned steps. For more infomation refer to this [tutorial](https://sdk.operatorframework.io/docs/building-operators/golang/tutorial/)
- Generate the manifests using `config/default`
- Comment out `ClusterRole` and `ClusterRoleBinding`. 
- Create a `ClusterRole` template in a `ScopeTemplate` and associate it with a `ScopeInstance`
- Run the operator, the `memcached-operator` will start complaining about `RBAC`.
- Run the `oria-operator`, it will create `ClusterRole` and `ClusterRoleBinding`. This will make sure that `memcached-operator` will not complain about `RBAC`. 

### Install and run oria-operator

Clone the [github repo](https://github.com/operator-framework/oria-operator.git) and  run the below commands on the root of the directory.

```
$ make install

$ make run

> ```

Reviewer: To make it easy for people following along with the tutorial we should add the command needed to clone the repo. We can also combine the `make install` and `make run` commands into a single command (`make install run`). I would also put each command in it's own code block without the `$` so that they can copy the command easily and just paste it into the terminal.

I think it might also be worth adding a note at the end of this section that this will be a process that runs in the foreground and that they will need to open a new terminal window/tab to continue with the rest of the tutorial.

Author: The above process runs in the foreground. You will need to open a new terminal window/tab to continue with the rest of the tutorial.

Author: updated

### Create a new project

Use the [Operator SDK](https://sdk.operatorframework.io/docs/installation/) CLI to create a new memcached-operator from the [tutorial](https://sdk.operatorframework.io/docs/building-operators/golang/tutorial/).

```sh
mkdir -p $HOME/projects/memcached-operator
cd $HOME/projects/memcached-operator
# we'll use a domain of example.com
# so all API groups will be <group>.example.com

> operator-sdk init --domain example.com --repo github.com/example/memcached-operator

Reviewer: I would break all of these commands into their own code block for easier copy-paste. 

Alternatively we could combine all of these into one "command" like:
```
mkdir -p $HOME/projects/memcached-operator && \
cd $HOME/projects/memcached-operator && \
operator-sdk init --domain example.com --repo github.com/example/memcached-operator
```

Author: I am okay with the above approach but not sure if every user wants to create a directory at `$HOME/projects/memcached-operator`.

Sometimes I create it somewhere else. WDYT @everettraven ?

Reviewer: I think we can keep them separate commands if we wanted and just have the docs be like:

Create a directory for the memcached operator:
```
mkdir -p <directory>
```
Navigate into the newly created directory and initialize the project with:
```
operator-sdk init --domain example.com --repo github.com/example/memcached-operator
```

Author: yeah, that sounds good to me.

Author: updated

```

### Create a new API and Controller 

```sh
$ operator-sdk create api --group cache --version v1alpha1 --kind Memcached --resource --controller
Writing scaffold for you to edit...
api/v1alpha1/memcached_types.go
controllers/memcached_controller.go
...

> ```

Reviewer: I would remove the sample output from this code block and only include the command

Author: I followed the same tutorial and wanted to sync it whatever we have with golang tutorial just to avoid confusion.

Author: removed it

Reviewer: I understand trying to keep it in sync with the golang tutorial, but I think for this case we are trying not to reiterate more than necessary since ideally they will use the SDK golang tutorial for more detailed information. 

I am fine with leaving it this way if you feel it is best, but in a scenario like this I personally prefer providing only the information they need to get to the state we want them in

Author: agree and I removed that extra content.

```
// MemcachedSpec defines the desired state of Memcached
type MemcachedSpec struct {
	//+kubebuilder:validation:Minimum=0
	// Size is the size of the memcached deployment
	Size int32 `json:"size"`
}

// MemcachedStatus defines the observed state of Memcached
type MemcachedStatus struct {
	// Nodes are the names of the memcached pods
	Nodes []string `json:"nodes"`
}
```

### Now create docker image and push to the dockerhub

Your Makefile composes image tags either from values written at project initialization or from the CLI. In particular, IMAGE_TAG_BASE lets you define a common image registry, namespace, and partial name for all your image tags. Update this to another registry and/or namespace if the current value is incorrect. Afterwards you can update the IMG variable definition like so:

```
-IMG ?= controller:latest
+IMG ?= $(IMAGE_TAG_BASE):$(VERSION)
```

Once done, you do not have to set IMG or any other image variable in the CLI. The following command will build and push an operator image tagged as example.com/memcached-operator:v0.0.1 to Docker Hub:

```

> make docker-build docker-push

Reviewer: I would update this to include the `IMG` environment variable so people following the tutorial can replace it with a value that will actually work.
```suggestion
make docker-build docker-push IMG=<registry>/<user>/<img-name>:<version>
```

Author: updated

```

### Generating CRD manifests 

```sh
$ mkdir manifest

> $ kustomize build config/default > manifest/manifest.yaml

Reviewer: 1. I would split this into 2 code blocks or combine the commands with `&&`
2. Where do I get `kustomize` from? (`make kustomize` might be a nice step to include here because it will download it to `bin/kustomize`)

Author: yeah, I will add as a prerequisite at the top.

kind: ScopeTemplate
metadata:
  name: scopetemplate-sample-manager
spec:
  clusterRoles:
  - generateName: manager-role
    rules:
    - apiGroups:
      - "*"
      resources:
      - deployments
      - memcacheds
      verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
    - apiGroups:
      - cache.example.com
      resources:
      - memcacheds/finalizers
      verbs:
      - update
    - apiGroups:
      - cache.example.com
      resources:
      - memcacheds/status
      verbs:
      - get
      - patch
      - update
    subjects:
    - kind: ServiceAccount
      name: memcached-operator-controller-manager
      namespace: memcached-operator-system
```

> Before this install and run `oria-operator`.

Reviewer: If these steps need to be done before running the above step, we should move them to be before it

Author: what do you mean? Shall I run `oria-operator` first and the start with other stuff?

Author: Updated

      - memcacheds/status
      verbs:
      - get
      - patch
      - update
    subjects:
    - kind: ServiceAccount
      name: memcached-operator-controller-manager
      namespace: memcached-operator-system
```

Then, apply manifest and look for the pod logs.

```
$ kubectl apply -f manifest
namespace/memcached-operator-system created
customresourcedefinition.apiextensions.k8s.io/memcacheds.cache.example.com created
serviceaccount/memcached-operator-controller-manager created

> role.rbac.authorization.k8s.io/memcached-operator-leader-election-role created

Reviewer: I would remove the sample output here and just leave the command

Author: removed

...
```

The `memcached-operator` is running successfully with some errors. Check the pod logs and now you will be able to see that pod is complaining about `RBAC`.

```
$ kubectl get pods -n memcached-operator-system
NAME                                                     READY   STATUS    RESTARTS   AGE
memcached-operator-controller-manager-684554467c-4jflr   2/2     Running   0          9m45s

> $ kubectl logs pod memcached-operator-controller-manager-684554467c-4jflr manager -n memcached-operator-system

Reviewer: I _think_ this can just be:
```suggestion
$ kubectl logs memcached-operator-controller-manager-684554467c-4jflr -n memcached-operator-system
```

Author: this is not showing actual logs if we don't pass manager. I will remove pod.

> ```

Reviewer: I would split this into separate code blocks

cat << EOF | kubectl apply -f -
apiVersion: operators.io.operator-framework/v1alpha1
kind: ScopeInstance
metadata:
  name: scopeinstance-sample-manager
spec:
  scopeTemplateName: scopetemplate-sample-manager
EOF
```

```
$ kubectl apply -f operators_v1_scopeinstance.yaml

> ```

Reviewer: You should be able to remove this command since the one above it handles it.

Author: definately

> The `oria-operator` will create the `ClusterRole` and `ClusterRoleBinding`. Then, the `memcached-operator` will pick these RBACs and now it will stop complaining.

Reviewer: I think it might be nice to add steps to show that the `ClusterRole` and the `ClusterRoleBinding` is created. Could just be simple `kubectl get ...` commands.

# Oria Operator Tutorial

> ### An in-depth walkthrough of building and running a oria-operator with memcached-operator.

Reviewer 2: I would just make this a paragraph after the title instead of a subsection

```suggestion
An in-depth walkthrough of building and running a oria-operator with memcached-operator.
```

## Prerequisites

> - [Operator SDK](https://sdk.operatorframework.io/docs/installation/) v1.8.0 or newer

Reviewer 2: ```suggestion
- [Operator SDK](https://sdk.operatorframework.io/docs/installation/) v1.21.0 or newer
```

Probably best to just pick a newer one to avoid hitting any issues with older versions.

Author: agree

> - User authorized with `cluster-admin` permissions.

Reviewer 2: I would probably add a reason why they should have cluster-admin. Something like "... permissions so that you can install an operator, CRDs and change rbac". 

That will give the user an idea of why. And they might be really advanced and choose to do it without clusteradmin because they have enough privileges based.

Author: updated

- [GNU Make](https://www.gnu.org/software/make/)

## Overview

We will create a sample project to let you know how it works and this sample will:

- Create a Memcached-Operator Deployment using this [tutorial](https://sdk.operatorframework.io/docs/building-operators/golang/tutorial/)
- Generate the manifests using `config/default`
- Run the Operator and after success, delete the `ClusterRole` and `ClusterRoleBinding`. Now, the `memcached-operator` will start complaining about it.

> - Create similar ClusterRole template in ScopeTemplate and assciate it with ScopeInstance

Reviewer 2: ```suggestion
- Create similar ClusterRole template in a ScopeTemplate and associate it with a ScopeInstance
```

> - Run the `oria-operator`, it will create `ClusterRole` and `ClusterRoleBinding`. This will make sure that `memcached-operator` will not complain about RBAC's. 

Author: agree. will add scopeInstance at the end.

### Create a new project

> Use the [Operator SDK](https://sdk.operatorframework.io/docs/installation/) CLI to create a new memcached-operator from the [tutorial](https://sdk.operatorframework.io/docs/building-operators/golang/tutorial/).

Reviewer 2: We should mention that they don't have to read the entire tutorial from SDK. Only to reference it for more details, but below we will be showing them what they need for the purposes of this demo. 

Not quite sure how to rephrase it, but as it reads now it seems like they MUST follow the SDK tutorial first BEFORE doing anything on the oria side. I don't think that's what you intended.

Author: Yes, I added all the steps and pointed SDK tutorial just for information.

Author: updated.

cd $HOME/projects/memcached-operator
# we'll use a domain of example.com
# so all API groups will be <group>.example.com
operator-sdk init --domain example.com --repo github.com/example/memcached-operator
```

### Create a new API and Controller 

```sh
$ operator-sdk create api --group cache --version v1alpha1 --kind Memcached --resource --controller
Writing scaffold for you to edit...
api/v1alpha1/memcached_types.go
controllers/memcached_controller.go
...
```

Define the API for the Memcached Custom Resource(CR) by modifying the Go type definitions at `api/v1alpha1/memcached_types.go` to have the following spec and status:

```
// MemcachedSpec defines the desired state of Memcached
type MemcachedSpec struct {
	//+kubebuilder:validation:Minimum=0
	// Size is the size of the memcached deployment
	Size int32 `json:"size"`
}

// MemcachedStatus defines the observed state of Memcached
type MemcachedStatus struct {
	// Nodes are the names of the memcached pods
	Nodes []string `json:"nodes"`
}
```

### Now create docker image and push to the dockerhub

Change below code in Makefile

```
-IMG ?= controller:latest
+IMG ?= $(IMAGE_TAG_BASE):$(VERSION)
```

> Once done, you do not have to set IMG or any other image variable in the CLI. The following command will build and push an operator image tagged as example.com/memcached-operator:v0.0.1 to Docker Hub:

Reviewer 2: What do we want them to set the IMAGE_TAG_BASE and VERSION to? Seems like they should pass something in for that. I *think* you're suggesting that they use `example.com/memcached-operator:v0.0.1` but this won't work as written. 

There needs to be a step that shows the user how to set `IMAGE_TAG_BASE` and `VERSION` to the above values.

Author: added content at the top about `IMAGE_TAG_BASE`

```

> make docker-build docker-push

Reviewer 2: Yeah without the instructions for setting `IMAGE_TAG_BASE` and `VERSION` this `docker-*` command will push to empty.

Author: Added comment at top same like go-lang tutorial

```

### Generating CRD manifests 

```sh
$ kustomize build config/default > manifest/manifest.yaml
```

> Need to update `manifest.yaml` file.

Reviewer 2: I'd make this another subsection

```suggestion
### Update the `manifest.yaml` file
```

Author: yes

Update image tag from `controller:latest` to `example.com/memcached-operator:v0.0.1`

> Then, remove `ClusterRole` with name `memcached-operator-manager-role` and `ClsuerRoleBinding` with name `memcached-operator-manager-rolebinding`.

Reviewer 2: typo
```suggestion
Then, remove `ClusterRole` with name `memcached-operator-manager-role` and `ClusterRoleBinding` with name `memcached-operator-manager-rolebinding`.
```

> Now, create the `ScopeTemplate` and add above removed `ClsuterRole` in spec in `manifests/scopeTemplate.yaml`.

Reviewer 2: ```suggestion
Now, create the `ScopeTemplate` and add above removed `ClusterRole` in spec in `manifests/scopeTemplate.yaml`.
```

```yaml
apiVersion: operators.io.operator-framework/v1alpha1
kind: ScopeTemplate
metadata:
  name: scopetemplate-sample-manager
spec:
  clusterRoles:
  - generateName: manager-role
    rules:
    - apiGroups:
      - "*"
      resources:

>       - "*"

Author: sure

      verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
    - apiGroups:
      - cache.example.com
      resources:
      - memcacheds/finalizers
      verbs:
      - update
    - apiGroups:
      - cache.example.com
      resources:
      - memcacheds/status
      verbs:
      - get
      - patch
      - update
    subjects:
    - kind: ServiceAccount
      name: memcached-operator-controller-manager
      namespace: memcached-operator-system
```

Before this install `ScopeTemplate` CRD from `oria-operator`.

```

> $ make install

Reviewer 2: Where are they running this `make install` command? Won't that install the memcached CRDs? I presume I'm still in the memcached-operator directory

Author: Yeah, User need to diver to `oria-operator` and do this as we are running `ScopeTemplate` from manifests.

```
$ kubectl apply -f manifest
namespace/memcached-operator-system created
customresourcedefinition.apiextensions.k8s.io/memcacheds.cache.example.com created
serviceaccount/memcached-operator-controller-manager created
role.rbac.authorization.k8s.io/memcached-operator-leader-election-role created
...
```

The `memcached-operator` is running successfully with some errors. Check the pod logs and now you will be able to see that pod is complaining about `RBAC`.

```
$ kubectl get pods -n memcached-operator-system
NAME                                                     READY   STATUS    RESTARTS   AGE
memcached-operator-controller-manager-684554467c-4jflr   2/2     Running   0          9m45s

$ kubectl describe pod memcached-operator-controller-manager-684554467c-4jflr manager -n memcached-operator-system
```

Now, run the `oria-operator`

```
$ make run
```

Create `ScopeInstance` that refers to `ScopeTemplate` and apply it.

```yaml
apiVersion: operators.io.operator-framework/v1alpha1
kind: ScopeInstance
metadata:
  name: scopeinstance-sample-manager
spec:
  scopeTemplateName: scopetemplate-sample-manager
```

```
$ kubectl apply -f operators_v1_scopeinstance.yaml

> ```

Reviewer: There is a handy shortcut for tutorials/demos that I like to use when mentioning that a smaller manifest should be applied. Instead of having them create the YAML file and apply it by hand I normally provide a command that will apply it for them. In this case I would set the command as:
```
cat << EOF | kubectl apply -f -
apiVersion: operators.io.operator-framework/v1alpha1
kind: ScopeInstance
metadata:
  name: scopeinstance-sample-manager
spec:
  scopeTemplateName: scopetemplate-sample-manager
EOF
```
This way they can copy and paste the command and follow along easier than having to create a bunch of files.

We don't have to do this in this tutorial, but I just figured I would share since it is how I normally create mine :)

Author: updated above content as like above.

## Prerequisites

- [Operator SDK](https://sdk.operatorframework.io/docs/installation/) v1.21.0 or newer
- User authorized with `cluster-admin` permissions so that you can install an operator, CRDs and change rbac.
- [GNU Make](https://www.gnu.org/software/make/)
- Create an [Kind Cluster](https://kind.sigs.k8s.io/)
- Install [kustomize](https://kubectl.docs.kubernetes.io/installation/kustomize/)

## Overview

We will create a sample project to let you know how it works and this sample will:

- Create a `memcached-operator` using below mentioned steps. For more infomation refer to this [tutorial](https://sdk.operatorframework.io/docs/building-operators/golang/tutorial/)
- Generate the manifests using `config/default`
- Comment out `ClusterRole` and `ClusterRoleBinding`. 
- Create a `ClusterRole` template in a `ScopeTemplate` and associate it with a `ScopeInstance`
- Run the operator, the `memcached-operator` will start complaining about `RBAC`.
- Run the `oria-operator`, it will create `ClusterRole` and `ClusterRoleBinding`. This will make sure that `memcached-operator` will not complain about `RBAC`. 

### Install and run oria-operator

> Clone the [github repo](https://github.com/operator-framework/oria-operator.git) and  run the below commands on the root of the directory.

Reviewer: I feel like we should provide them the clone command like:
```
git clone https://github.com/operator-framework/oria-operator.git
```

Author: Added

# Oria Operator Tutorial

An in-depth walkthrough of building and running a oria-operator with memcached-operator.

## Prerequisites

- [Operator SDK](https://sdk.operatorframework.io/docs/installation/) v1.21.0 or newer
- User authorized with `cluster-admin` permissions so that you can install an operator, CRDs and change rbac.
- [GNU Make](https://www.gnu.org/software/make/)
- Create an [Kind Cluster](https://kind.sigs.k8s.io/)
- Install [kustomize](https://kubectl.docs.kubernetes.io/installation/kustomize/)

## Overview

We will create a sample project to let you know how it works and this sample will:

- Create a `memcached-operator` using below mentioned steps. For more infomation refer to this [tutorial](https://sdk.operatorframework.io/docs/building-operators/golang/tutorial/)
- Generate the manifests using `config/default`
- Comment out `ClusterRole` and `ClusterRoleBinding`. 
- Create a `ClusterRole` template in a `ScopeTemplate` and associate it with a `ScopeInstance`
- Run the operator, the `memcached-operator` will start complaining about `RBAC`.

> - Run the `oria-operator`, it will create `ClusterRole` and `ClusterRoleBinding`. This will make sure that `memcached-operator` will not complain about `RBAC`. 

Reviewer 2: I like this summary. Gives the user a good idea of what they are about to do.

Author: cool

### Install and run oria-operator

Clone the [github repo](https://github.com/operator-framework/oria-operator.git)

```
git clone https://github.com/operator-framework/oria-operator.git
```

Then, run the `oria-operator` with below commands on the root of the directory.

```

> $ make install run

Reviewer: ```suggestion
make install run
```

Author: done

```

The above process runs in the foreground. You will need to open a new terminal window/tab to continue with the rest of the tutorial.

### Create a new project

Use the [Operator SDK](https://sdk.operatorframework.io/docs/installation/) CLI to create a new memcached-operator from the [tutorial](https://sdk.operatorframework.io/docs/building-operators/golang/tutorial/).

Create a directory for the memcached operator:

```
mkdir -p <directory>
```

Navigate into the newly created directory and initialize the project with:

```sh
operator-sdk init --domain example.com --repo github.com/example/memcached-operator
```

### Create a new API and Controller 

```sh

> $ operator-sdk create api --group cache --version v1alpha1 --kind Memcached --resource --controller

Reviewer: ```suggestion
operator-sdk create api --group cache --version v1alpha1 --kind Memcached --resource --controller
```

```

Define the API for the Memcached Custom Resource(CR) by modifying the Go type definitions at `api/v1alpha1/memcached_types.go` to have the following spec and status:

```
// MemcachedSpec defines the desired state of Memcached
type MemcachedSpec struct {
	//+kubebuilder:validation:Minimum=0
	// Size is the size of the memcached deployment
	Size int32 `json:"size"`
}

// MemcachedStatus defines the observed state of Memcached
type MemcachedStatus struct {
	// Nodes are the names of the memcached pods
	Nodes []string `json:"nodes"`
}
```

### Now create docker image and push to the dockerhub

Your Makefile composes image tags either from values written at project initialization or from the CLI. In particular, IMAGE_TAG_BASE lets you define a common image registry, namespace, and partial name for all your image tags. Update this to another registry and/or namespace if the current value is incorrect. Afterwards you can update the IMG variable definition like so:

```
-IMG ?= controller:latest
+IMG ?= $(IMAGE_TAG_BASE):$(VERSION)

> ```

Reviewer 2: A follow-on PR, we can remove this because we are overriding the `IMG` value with the command below. That is perfectly acceptable. Keep it here for now.

Author: updated this too.

Once done, you do not have to set IMG or any other image variable in the CLI. The following command will build and push an operator image tagged as example.com/memcached-operator:v0.0.1 to Docker Hub:

```

> make docker-build docker-push IMG=<registry>/<user>/<img-name>:<version>

Reviewer 2: +1 this is good.

Author: sure

```

### Generating CRD manifests 

```sh

> $ mkdir manifest && \ 

Reviewer: ```suggestion
mkdir manifest && \ 
```

Update image tag from `controller:latest` to `example.com/memcached-operator:v0.0.1`

Then, remove `ClusterRole` with name `memcached-operator-manager-role` and `ClusterRoleBinding` with name `memcached-operator-manager-rolebinding`.

Now, create the `ScopeTemplate` and add above removed `ClusterRole` in spec in `manifests/scopeTemplate.yaml`.

```yaml
apiVersion: operators.io.operator-framework/v1alpha1
kind: ScopeTemplate
metadata:
  name: scopetemplate-sample-manager
spec:
  clusterRoles:
  - generateName: manager-role
    rules:
    - apiGroups:
      - "*"
      resources:
      - deployments
      - memcacheds
      verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
    - apiGroups:
      - cache.example.com
      resources:
      - memcacheds/finalizers
      verbs:
      - update
    - apiGroups:
      - cache.example.com
      resources:
      - memcacheds/status
      verbs:
      - get
      - patch
      - update
    subjects:
    - kind: ServiceAccount
      name: memcached-operator-controller-manager
      namespace: memcached-operator-system
```

Then, apply manifest and look for the pod logs.

```

> $ kubectl apply -f manifest

Reviewer: ```suggestion
kubectl apply -f manifest
```

```

The `memcached-operator` is running successfully with some errors. 

Get the list of pods running under `memcached-operator-system` namespace.

```
$ kubectl get pods -n memcached-operator-system
NAME                                                     READY   STATUS    RESTARTS   AGE
memcached-operator-controller-manager-684554467c-4jflr   2/2     Running   0          9m45s

> ```

Reviewer: I would change this to something more like:

---

```
kubectl get pods -n memcached-operator-system
```
The output should be similar to
```
NAME                                                     READY   STATUS    RESTARTS   AGE
memcached-operator-controller-manager-684554467c-4jflr   2/2     Running   0          9m45s
```
---

Check the pod logs and now you will be able to see that pod is complaining about `RBAC`.

```

> $ kubectl logs memcached-operator-controller-manager-684554467c-4jflr manager -n memcached-operator-system

Reviewer: ```suggestion
kubectl logs memcached-operator-controller-manager-684554467c-4jflr manager -n memcached-operator-system
```

```

Create `ScopeInstance` that refers to `ScopeTemplate` and apply it.

```
cat << EOF | kubectl apply -f -
apiVersion: operators.io.operator-framework/v1alpha1
kind: ScopeInstance
metadata:
  name: scopeinstance-sample-manager
spec:
  scopeTemplateName: scopetemplate-sample-manager
EOF
```

The `oria-operator` will create the `ClusterRole` and `ClusterRoleBinding`. Verify `ClusterRole` and `ClusterRoleBinding` with below commands.

```
$ kubectl get clusterroles

$ kubectl get clusterrolebinding

> ```

Reviewer: I would change this to be more like:

---

```
kubectl get clusterroles
```

In the list you should see a `ClusterRole` with the name ...

```
kubectl get clusterrolebindings
```

In the list you should see a `ClusterRoleBinding` with the name ...

---
